File: smart_trip_ai_mobile/lib/domain/models/telemetry.dart
enum DrivingState {
  driving,
  idling,
  pitStop,
  offline
}

class TelemetryPoint {
  final String id;
  final String tripId;
  final int timestamp; // epoch ms
  final double latitude;
  final double longitude;
  final double? altitude;
  final double speedKmH;
  final double? heading; // 0-360 deg
  final double? accuracy; // meters

  // 3-axis Accelerometer (m/s^2)
  final double accelX; // Lateral (left/right)
  final double accelY; // Vertical (up/down)
  final double accelZ; // Longitudinal (forward/backward)

  // Gyroscope (deg/s)
  final double gyroX; // Pitch
  final double gyroY; // Roll
  final double gyroZ; // Yaw (turn rate)

  // Derived state
  final DrivingState state;
  final double gForceCombined; // Combined G-Force magnitude
  final String? safetyPenalty; // 'HARD_ACCEL' | 'HARD_BRAKE' | 'SWERVE' | null

  TelemetryPoint({
    required this.id,
    required this.tripId,
    required this.timestamp,
    required this.latitude,
    required this.longitude,
    this.altitude,
    required this.speedKmH,
    this.heading,
    this.accuracy,
    required this.accelX,
    required this.accelY,
    required this.accelZ,
    required this.gyroX,
    required this.gyroY,
    required this.gyroZ,
    required this.state,
    required this.gForceCombined,
    this.safetyPenalty,
  });
}

enum PitStopCategory {
  gasStation,
  coffeeShop,
  restArea,
  parking,
  clientSite,
  trafficHold
}

class PitStop {
  final String id;
  final String tripId;
  final int startTime;
  final int? endTime;
  final int durationSeconds;
  final double latitude;
  final double longitude;
  final String? address;
  final PitStopCategory category;
  final String name;
  final double idleFuelCost;
  final double idleFuelLiters;

  PitStop({
    required this.id,
    required this.tripId,
    required this.startTime,
    this.endTime,
    required this.durationSeconds,
    required this.latitude,
    required this.longitude,
    this.address,
    required this.category,
    required this.name,
    required this.idleFuelCost,
    required this.idleFuelLiters,
  });
}

class DrivingScore {
  final double safetyScore; // 0 - 100
  final double ecoScore;    // 0 - 100
  final double overallScore;// 0 - 100
  final int hardAccelerationsCount;
  final int hardBrakesCount;
  final int sharpSwervesCount;
  final int idleTimeSeconds;
  final double smoothDrivingDistanceKm;

  DrivingScore({
    required this.safetyScore,
    required this.ecoScore,
    required this.overallScore,
    required this.hardAccelerationsCount,
    required this.hardBrakesCount,
    required this.sharpSwervesCount,
    required this.idleTimeSeconds,
    required this.smoothDrivingDistanceKm,
  });
}

enum VehicleType { sedan, suv, truck, hybrid, ev }
enum FuelType { gasoline, diesel, hybrid, electric }

class VehicleConfig {
  final String vehicleName;
  final VehicleType vehicleType;
  final FuelType fuelType;
  final double engineSizeLiters; // e.g. 2.0
  final double idleConsumptionRateLph; // L/hour idle (e.g. 1.2 L/h)
  final double fuelPricePerLiter; // e.g. 1.45 $/L

  VehicleConfig({
    required this.vehicleName,
    required this.vehicleType,
    required this.fuelType,
    required this.engineSizeLiters,
    required this.idleConsumptionRateLph,
    required this.fuelPricePerLiter,
  });
}

enum TripStatus { active, completed }

class Trip {
  final String id;
  final String title;
  final int startTime;
  final int? endTime;
  final int durationSeconds;
  final double distanceKm;
  final double maxSpeedKmH;
  final double avgSpeedKmH;
  
  // Fuel & Cost Metrics
  final double totalFuelLiters;
  final double idleFuelLiters;
  final double totalCost;
  final double idleCost;
  final double costSavedCruising;

  // Driving & Pedestrian Metrics
  final double safetyScore;
  final double ecoScore;
  final int? stepsCount;
  final double? caloriesBurned;
  
  // Event Counters
  final int hardAccelerations;
  final int hardBrakes;
  final int sharpSwerves;
  
  // Polylines & Stops
  final List<TelemetryPoint> telemetryPoints;
  final List<PitStop> pitStops;
  final TripStatus status;

  Trip({
    required this.id,
    required this.title,
    required this.startTime,
    this.endTime,
    required this.durationSeconds,
    required this.distanceKm,
    required this.maxSpeedKmH,
    required this.avgSpeedKmH,
    required this.totalFuelLiters,
    required this.idleFuelLiters,
    required this.totalCost,
    required this.idleCost,
    required this.costSavedCruising,
    required this.safetyScore,
    required this.ecoScore,
    this.stepsCount,
    this.caloriesBurned,
    required this.hardAccelerations,
    required this.hardBrakes,
    required this.sharpSwerves,
    required this.telemetryPoints,
    required this.pitStops,
    required this.status,
  });
}

enum AIVisualizationType { card, table, barChart, pieChart }

class AIQueryResponse {
  final String query;
  final String generatedSQL;
  final String answerSummary;
  final List<dynamic> data;
  final AIVisualizationType? visualizationType;
  final int timestamp;

  AIQueryResponse({
    required this.query,
    required this.generatedSQL,
    required this.answerSummary,
    required this.data,
    this.visualizationType,
    required this.timestamp,
  });
}

File: src/domain/models/telemetry.ts
export type DrivingState = 'DRIVING' | 'IDLING' | 'PIT_STOP' | 'OFFLINE';

export interface TelemetryPoint {
  id: string;
  tripId: string;
  timestamp: number; // epoch ms
  latitude: number;
  longitude: number;
  altitude?: number;
  speedKmH: number;
  heading?: number; // 0-360 deg
  accuracy?: number; // meters

  // 3-axis Accelerometer (m/s^2)
  accelX: number; // Lateral (left/right)
  accelY: number; // Vertical (up/down)
  accelZ: number; // Longitudinal (forward/backward)

  // Gyroscope (deg/s)
  gyroX: number; // Pitch
  gyroY: number; // Roll
  gyroZ: number; // Yaw (turn rate)

  // Derived state
  state: DrivingState;
  gForceCombined: number; // Combined G-Force magnitude
  safetyPenalty?: string; // 'HARD_ACCEL' | 'HARD_BRAKE' | 'SWERVE' | null
}

export interface PitStop {
  id: string;
  tripId: string;
  startTime: number;
  endTime?: number;
  durationSeconds: number;
  latitude: number;
  longitude: number;
  address?: string;
  category: 'GAS_STATION' | 'COFFEE_SHOP' | 'REST_AREA' | 'PARKING' | 'CLIENT_SITE' | 'TRAFFIC_HOLD';
  name: string;
  idleFuelCost: number;
  idleFuelLiters: number;
}

export interface DrivingScore {
  safetyScore: number; // 0 - 100
  ecoScore: number;    // 0 - 100
  overallScore: number;// 0 - 100
  hardAccelerationsCount: number;
  hardBrakesCount: number;
  sharpSwervesCount: number;
  idleTimeSeconds: number;
  smoothDrivingDistanceKm: number;
}

export interface VehicleConfig {
  vehicleName: string;
  vehicleType: 'SEDAN' | 'SUV' | 'TRUCK' | 'HYBRID' | 'EV';
  fuelType: 'GASOLINE' | 'DIESEL' | 'HYBRID' | 'ELECTRIC';
  engineSizeLiters: number; // e.g. 2.0
  idleConsumptionRateLph: number; // L/hour idle (e.g. 1.2 L/h)
  fuelPricePerLiter: number; // e.g. 1.45 $/L
}

export interface Trip {
  id: string;
  title: string;
  startTime: number;
  endTime?: number;
  durationSeconds: number;
  distanceKm: number;
  maxSpeedKmH: number;
  avgSpeedKmH: number;
  
  // Fuel & Cost Metrics
  totalFuelLiters: number;
  idleFuelLiters: number;
  totalCost: number;
  idleCost: number;
  costSavedCruising: number;

  // Driving & Pedestrian Metrics
  safetyScore: number;
  ecoScore: number;
  stepsCount?: number;
  caloriesBurned?: number;
  
  // Event Counters
  hardAccelerations: number;
  hardBrakes: number;
  sharpSwerves: number;
  
  // Polylines & Stops
  telemetryPoints: TelemetryPoint[];
  pitStops: PitStop[];
  status: 'ACTIVE' | 'COMPLETED';
}

export interface AIQueryResponse {
  query: string;
  generatedSQL: string;
  answerSummary: string;
  data: any[];
  visualizationType?: 'CARD' | 'TABLE' | 'BAR_CHART' | 'PIE_CHART';
  timestamp: number;
}
